### ISSUE TITLE
Upstream :Lsp commands from nvim-lspconfig

### ISSUE THREAD

**Issue by gpanders** [neovim/neovim#28479]
This is a tracking issue to either upstream the `:LspStart`, `:LspStop`, `:LspRestart`, and `:LspInfo` commands from nvim-lspconfig or determine formally if those features are not wanted.

`:checkhealth lsp` does a lot of what `:LspInfo` does, but is not a complete replacement. In https://github.com/neovim/neovim/pull/18476#issuecomment-1120390374, @mfussenegger mentioned:

>:lsp / :LspInfo: LspInfo currently does more than just printing all clients, it also prints all not started yet. It can do that because lsp-config currently includes the "managing clients" functionality. I think the first step we have to do to bring anything like that into core is to move the manager component.

Nvim core would need the concept of "registering" or "managing" a client without it being started. For example, a `vim.lsp.register` API that makes Nvim aware of a client configuration without actually starting it. This same functionality would be needed for `:LspStart` as well, if we want to include that.

I do not think we should include `:LspInfo` directly, but its functionality should be integrated into `:checkhealth lsp`.

If we do add `:LspStart` and/or `:LspStop`, I propose they be added as `:lsp start` and `:lsp stop` and promote `:lsp` to a first class command (note, this is a different interface thn the one proposed in #18476).

cc @mfussenegger

**Comment by justinmk** [comment]
> Nvim core would need the concept of "registering" or "managing" a client without it being started. For example, a `vim.lsp.register` API that makes Nvim aware of a client configuration without actually starting it.

Do we only need a way to "define" clients and add them to a list, or is there more involved (such as lifecycle, cleanup, ... which is the usual unpleasantness implied by a "registry")? Maybe `vim.lsp.add_client()` or `vim.lsp.def_client()` could be the name.

**Comment by lewis6991** [comment]
+1, but I would prefer we add `:Lsp` with `stop`, `restart` and `info` sub-commands.

**Comment by lewis6991** [comment]
> Nvim core would need the concept of "registering" or "managing" a client without it being started. For example, a vim.lsp.register API that makes Nvim aware of a client configuration without actually starting it. This same functionality would be needed for :LspStart as well, if we want to include that.

Note we already have `vim.treesitter.language.register/add` which serves a similar purpose.

**Comment by justinmk** [comment]
> we already have `vim.treesitter.language.register/add` which serves a similar purpose, would be _nice_ if we could mirror that, or have them align.

Definitely. gpanders mentioned `vim.lsp.config`, I wonder if that can be the interface for "defining a client". See also https://github.com/neovim/neovim/issues/24054

**Comment by lewis6991** [comment]
One benefit (and design requirement) of `vim.treesitter.language.add` is that it can be called from multiple sources. I think that requirement should apply to LSP to, i.e. servers can be specified in multiple sources: config, plugins, etc.

**Comment by clason** [comment]
We _definitely_ want to support "language server plugins" as a first-class use case here.

**Comment by lewis6991** [comment]
Just to throw in another idea.

I've been debugging some LSP stuff and wondering whether something like this would also be useful:

```lua
vim.lsp.log.set_level(vim.lsp.log.levels.DEBUG)
-- vim.lsp.log.set_format_func(vim.inspect)

api.nvim_create_user_command('LspLog', function()
  local log_path = vim.lsp.get_log_path()
  vim.cmd.split(log_path)
end, {})
```

Something _like_ this could reduce friction with LSP issues, since atm, getting the log file is rather tedious, and can quickly solve certain kinds of issues. Just an idea. A similar treesitter analogue is `:InspectTree`.

With that said, the LSP logging does need some improvements because currently the log is shared between all servers and nvim sessions and grows indefinitely unless it is manually cleared.

**Comment by gpanders** [comment]
>Something _like_ this could reduce friction with LSP issues, since atm, getting the log file is rather tedious

Not opposed to it, but just to point this out: `:checkh lsp` prints the path to the log file and you can use `gf` or `<C-W>f` on that to open it. It's admittedly not the most straightforward, but it's not too bad in a pinch.

**Comment by justinmk** [comment]
> I do not think we should include `:LspInfo` directly, but its functionality should be integrated into `:checkhealth lsp`.

Done in https://github.com/neovim/nvim-lspconfig/pull/3339 .

**Comment by therealjasonkenney** [comment]
With the caveat that I'm not a contributor (atm), part of this project goal could be achieved incrementally by organizing configs like tree sitter does, make a repo for each configuration that provides the defaults and some useful utilities. I currently have been copying and modifying the configs defined in this repo for my own more specific configuration. 

Then to start an lsp with the lspconfig settings one can do:

```
local on_init = function(client)
  local client_capabilities = client.capabilities
  local cmp_capabilities = require("cmp_nvim_lsp").default_capabilities()

  client.capabilities =
    vim.tbl_deep_extend("force", client_capabilities, cmp_capabilities)
end

local default_config = require("lspconfig-lua").default_config()

vim.lsp.start(
  vim.tbl_deep_extend("force", default_config, {
    on_init = on_init
  })
)
```

^ Probably more low level than the average Neovim user, but you're planning on moving the other features into Neovim right?

Just a thought :)

**Comment by justinmk** [comment]
`vim.lsp.config` in Nvim 0.11 addresses that already, and isn't related to this issue.

**Comment by toupeira** [comment]
Is there a plan to support the old behaviour of `:LspStart` without arguments, which starts/enables all servers matching the current filetype? After https://github.com/neovim/nvim-lspconfig/pull/3734 the command now requires the server name to be passed on Neovim 0.11.2+

It's not too difficult to implement manually (see my snippet in https://github.com/neovim/neovim/discussions/33978#discussioncomment-13372488), but might be useful for others. I also currently loop through `vim.lsp.config._configs` to identify all configured LSPs and get their `filetypes`, so that might be a candidate for an API too (maybe just a getter `vim.lsp.get_configs()`, or something like `vim.lsp.get_configs_for_filetype(ft)`).

`LspStop` has the same change (it now requires an argument), but that one is more straightforward and could maybe be a `:lsp stop_all/disable_all` subcommand, which just disables all enabled servers.

(if any of this seems sensible I'd be curious to submit a PR 😀)

**Comment by shmerl** [comment]
> lsp stop_all/disable_all

I'd clarify that this should probably apply to all LSPs related to the filetype of the buffer your are disabling it in, not literally all configured LSPs, right? I.e. behavior I want is easy toggling of LSP for the specific buffer file type.

**Comment by toupeira** [comment]
@shmerl oh yes sorry, that's what I had in mind too. To avoid confusing naming and support both use-cases, maybe the signature could be `:lsp stop_all [filetype]`, where the filetype is optional.

**Comment by justinmk** [comment]
> support the old behaviour of `:LspStart` without arguments, which starts/enables all servers matching the current filetype? After [neovim/nvim-lspconfig#3734](https://github.com/neovim/nvim-lspconfig/pull/3734) the command now requires the server name to be passed on Neovim 0.11.2+

yes, please send a pr to https://github.com/neovim/nvim-lspconfig . that is the "staging area" until we get around to upstreaming this. (which isn't really blocked, except we likely want a subcommand interface like `:Lsp foo` instead of `:LspFoo`)

**Comment by toupeira** [comment]
@justinmk thanks! I submitted https://github.com/neovim/nvim-lspconfig/pull/3890 as a stopgap to restore the old behaviour.

**Comment by shmerl** [comment]
This will work for explicitly configured ones in vim.lsp.config, but for more final version in neovim itself it would be better to take in account also configurations `lsp/*.lua` though not sure how to make it differentiate all vs ones user might prefer ("configured"?).

**Comment by brianhuster** [comment]
> I've been debugging some LSP stuff and wondering whether something like this would also be useful:
> ```lua
> api.nvim_create_user_command('LspLog', function()
>   vim.cmd.split(vim.lsp.get_log_path())
> end, {})

Instead of a command, what about just making it an env var (e.g. `$NVIM_LSP_LOG`)? It only requires adding _a single line_ to Nvim codebase, but will offer a lot more flexibility (it automatically work with command-modifiers, external commands, terminal emulator). Also Nvim already has `$NVIM_LOG_FILE`

**Comment by justinmk** [comment]
`$NVIM_LOG_FILE` is not a pattern we're going to proliferate. Rather the plan is to add something like `:editlog` #35082

**Comment by ritchielrez** [comment]
I was wondering, if the `autostart` field in LSP config would be supported as they were before with `nvim-lspconfig`. It allowed users to disable auto start functionality of certain LSPs and then start them when they wanted with `:LspStart`.

**Comment by toupeira** [comment]
@ritchielrez I'm not sure `autostart` is still needed, you can achieve the same by simply not calling `vim.lsp.enable()` (`vim.lsp.config()` is fine though) and using `:LspStart` as before. Also see the discussion in https://github.com/neovim/neovim/discussions/33978

If you use `mason-lspconfig.nvim` you can set `automatic_enable = false` in its setup.

**Comment by superatomic** [comment]
For when this is upstreamed from lspconfig, `:LspStop` and `:LspStart` should really be named `:lsp disable` and `:lsp enable`, since they both wrap `vim.lsp.enable()`, and not `vim.lsp.stop_client()`. An additional `:lsp stop` could exist, but it shouldn't disable the client (i.e. the client gets restarted when the buffer is reloaded with `:edit`).

Edit: Opened https://github.com/neovim/nvim-lspconfig/issues/4173. It's not necessary to make this change before these commands are upstreamed, but it might still be worth doing.

**Update:** The issue was closed as to not break compatibility with users of nvim-lspconfig. I still strongly recommend changing the name during the upstreaming process for all of the reasons outlined in the issue, as we will not need to worry about compatibility at that point.
